cuML/src/solver/sgd.h

```diff
@@ -0,0 +1,267 @@
+		    ML::lr_type lr_type,
+		    math_t eta0,
+		    math_t power_t,
+		    ML::loss_funct loss,
+		    Functions::penalty penalty,
+		    math_t alpha,
+		    math_t l1_ratio,
+		    bool shuffle,
+		    math_t tol,
+		    int n_iter_no_change,
+		    cublasHandle_t cublas_handle,
+		    cusolverDnHandle_t cusolver_handle) {
+
+	ASSERT(n_cols > 1,
+			"Parameter n_cols: number of columns cannot be less than two");
+	ASSERT(n_rows > 1,
+			"Parameter n_rows: number of rows cannot be less than two");
+
+	math_t *mu_input = NULL;
+	math_t *mu_labels = NULL;
+	math_t *norm2_input = NULL;
+
+	if (fit_intercept) {
+		allocate(mu_input, n_cols);
+		allocate(mu_labels, 1);
+
+		GLM::preProcessData(input, n_rows, n_cols, labels, intercept, mu_input,
+				mu_labels, norm2_input, fit_intercept, false, cublas_handle,
+				cusolver_handle);
+	}
+
+	math_t *grads = NULL;
+	math_t *input_batch = NULL;
+	math_t *labels_batch = NULL;
+	math_t *loss_value = NULL;
+	int *indices = NULL;
+
+	allocate(grads, n_cols, true);
+	allocate(indices, batch_size);
+	allocate(input_batch, batch_size * n_cols);
+	allocate(labels_batch, batch_size);
+	allocate(loss_value, 1);
+
+	math_t prev_loss_value = math_t(0);
+	math_t curr_loss_value = math_t(0);
+
+	std::vector<int> rand_indices(n_rows);
+	std::mt19937 g(rand());
+	initShuffle(rand_indices, g);
+
+	math_t t = math_t(1);
+	math_t learning_rate = math_t(0);
+	if (lr_type == ML::lr_type::ADAPTIVE) {
+		learning_rate = eta0;
+	} else if (lr_type == ML::lr_type::OPTIMAL) {
+		eta0 = calOptimalInit(alpha);
+	}
+
+	int n_iter_no_change_curr = 0;
+
+	for (int i = 0; i < epochs; i++) {
+		int cbs = 0;
+		int j = 0;
+
+		if (i > 0 && shuffle) {
+			Solver::shuffle(rand_indices, g);
+		}
+
+		while (j < n_rows) {
+			if ((j + batch_size) > n_rows) {
+				cbs = n_rows - j;
+			} else {
+				cbs = batch_size;
+			}
+
+			if (cbs == 0)
+				break;
+
+			updateDevice(indices, &rand_indices[j], cbs);
+			Matrix::copyRows(input, n_rows, n_cols, input_batch, indices, cbs);
+			Matrix::copyRows(labels, n_rows, 1, labels_batch, indices, cbs);
+
+			if (loss == ML::loss_funct::SQRD_LOSS) {
+				Functions::linearRegLossGrads(input_batch, cbs, n_cols, labels_batch,
+						coef, grads, penalty, alpha, l1_ratio, cublas_handle);
+			} else if (loss == ML::loss_funct::LOG) {
+				Functions::logisticRegLossGrads(input_batch, cbs, n_cols, labels_batch,
+										coef, grads, penalty, alpha, l1_ratio, cublas_handle);
+			} else if (loss == ML::loss_funct::HINGE) {
+				Functions::hingeLossGrads(input_batch, cbs, n_cols, labels_batch,
+														coef, grads, penalty, alpha, l1_ratio, cublas_handle);
+			} else {
+				ASSERT(false,
+						"sgd.h: Other loss functions have not been implemented yet!");
+			}
+
+			if (lr_type != ML::lr_type::ADAPTIVE)
+			    learning_rate = calLearningRate(lr_type, eta0, power_t, alpha, t);
+
+			LinAlg::scalarMultiply(grads, grads, learning_rate, n_cols);
+			LinAlg::subtract(coef, coef, grads, n_cols);
```

Comment:
This could be fused into a single kernel call using something like
```
auto update = [=] __device__ (const math_t coef, const math_t grad){return coef - learning_rate * grad;};
binaryOp(coef, coef, grad, update);
```

Reply:
We are building the algorithms on top of ML primitives. If we create a separate kernel for everything, we cannot maintain the code easily.

Reply:
Actually, `binaryOp` is part of ml prims.
But I guess, this is perhaps not the most important thing to worry about at the moment.

Reply:
Good point. Will update it when we have time.